File: Week_10/Practicum/Course Schedule II.cpp
class Solution {
public:
    stack<int> topo;
    
    bool dfs(int curr, vector<vector<int>>& graph, vector<int>& state) {
        // state[curr]: 0=unvisited, 1=visiting, 2=visited
        state[curr] = 1;

        for (auto& neigh : graph[curr]) {
            if (state[neigh] == 1) {
                // Found a back edge -> cycle
                return true;
            } else if (state[neigh] == 0) {
                if (dfs(neigh, graph, state)) {
                    return true; // cycle found in deeper call
                }
            }
        }

        state[curr] = 2;
        topo.push(curr);
        return false;
    }

    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        for (auto &p : prerequisites) {
            int a = p[0];
            int b = p[1];
            graph[b].push_back(a);
        }

        vector<int> state(numCourses, 0);
        
        for (int i = 0; i < numCourses; i++) {
            if (state[i] == 0) {
                if (dfs(i, graph, state)) { // Cycle detected
                    return {};
                }
            }
        }

        vector<int> result;
        while (!topo.empty()) {
            result.push_back(topo.top());
            topo.pop();
        }
        return result;
    }
};

File: Week_10/Practicum/Course Schedule II.py
def dfs(graph, visited, current):
    visited[current] = 1

    for neighb in graph[current]:
        if visited[neighb] == 1:
            return True
        elif visited[neighb] == 0:
            if dfs(graph, visited, neighb):
                return True

    visited[current] = 2
    return False

def topo_sort(graph, in_degree, n):
    print(in_degree)
    q = deque()
    for node in range(n):
        if in_degree[node] == 0:
            q.append(node)

    ans = []

    while q:
        node = q.pop()
        ans.append(node)
        for neighb in graph[node]:
            in_degree[neighb] -= 1
            if in_degree[neighb] == 0:
                q.append(neighb)

    return ans

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        in_degree = defaultdict(int)

        for edge in prerequisites:
            graph[edge[1]].append(edge[0])
            in_degree[edge[0]] += 1

        visited = defaultdict(int)

        for node in list(graph.keys()):
            if visited[node] == 0 and dfs(graph, visited, node):
                return []

        return topo_sort(graph, in_degree, numCourses)
        